File: app/src/lib/modelThinking.ts
import type { AppSettings, ModelThinkingConfig, ModelThinkingLevel } from '@/types'

export const MODEL_THINKING_LEVELS = [
  'off',
  'minimal',
  'low',
  'medium',
  'high',
  'xhigh',
  'max',
] as const satisfies readonly ModelThinkingLevel[]

export interface ModelThinkingProfile {
  enabled: boolean
  levels: ModelThinkingLevel[]
  defaultLevel?: ModelThinkingLevel
  source: 'preset' | 'manual' | 'unavailable'
}

export const MODEL_THINKING_LEVEL_LABELS: Record<ModelThinkingLevel, string> = {
  off: '关闭',
  minimal: '最少',
  low: '低',
  medium: '中',
  high: '高',
  xhigh: '极高',
  max: '最大',
}

export function resolveModelThinking(
  settings: Pick<AppSettings, 'model_thinking'> | null | undefined,
  provider: string,
  model: string,
): ModelThinkingProfile {
  const manual = settings?.model_thinking?.[provider]?.[model]
  if (manual !== undefined) return profileFromConfig(manual, 'manual')
  const preset = builtInModelThinking(model)
  return preset
    ? profileFromConfig(preset, 'preset')
    : { enabled: false, levels: [], source: 'unavailable' }
}

export function effectiveModelThinkingLevel(
  profile: ModelThinkingProfile,
  requested: ModelThinkingLevel | string | null | undefined,
): ModelThinkingLevel | undefined {
  const normalized = normalizeThinkingLevel(requested)
  if (normalized && profile.levels.includes(normalized)) return normalized
  if (profile.defaultLevel && profile.levels.includes(profile.defaultLevel)) {
    return profile.defaultLevel
  }
  return profile.levels[0]
}

export function normalizeModelThinkingSettings(
  value: AppSettings['model_thinking'],
  providers: AppSettings['providers'],
): AppSettings['model_thinking'] {
  if (!value) return undefined
  const result: NonNullable<AppSettings['model_thinking']> = {}
  for (const [rawProvider, models] of Object.entries(value)) {
    const provider = rawProvider.trim()
    const configuredProvider = providers[provider]
    if (!provider || (provider !== 'tokenflux' && !configuredProvider?.custom)) continue
    const normalizedModels: Record<string, ModelThinkingConfig> = {}
    for (const [rawModel, configured] of Object.entries(models ?? {}).slice(0, 32)) {
      const model = rawModel.trim()
      if (
        !model
        || Array.from(model).length > 256
        || Array.from(model).some(character => {
          const code = character.charCodeAt(0)
          return code === 0 || code === 10 || code === 13
        })
      ) continue
      normalizedModels[model] = normalizeModelThinkingConfig(configured)
    }
    if (Object.keys(normalizedModels).length) result[provider] = normalizedModels
  }
  return Object.keys(result).length ? result : undefined
}

export function normalizeModelThinkingConfig(
  value: Partial<ModelThinkingConfig> | null | undefined,
): ModelThinkingConfig {
  const selected = new Set((value?.levels ?? []).flatMap(level => {
    const normalized = normalizeThinkingLevel(level)
    return normalized ? [normalized] : []
  }))
  const levels = MODEL_THINKING_LEVELS.filter(level => selected.has(level))
  const requestedDefault = normalizeThinkingLevel(value?.default_level)
  if (!value?.enabled) {
    return {
      enabled: false,
      levels,
      ...(levels.length
        ? { default_level: requestedDefault && levels.includes(requestedDefault)
            ? requestedDefault
            : preferredThinkingDefault(levels) }
        : {}),
    }
  }
  if (!levels.length) levels.push('low', 'medium', 'high')
  return {
    enabled: true,
    levels,
    default_level: requestedDefault && levels.includes(requestedDefault)
      ? requestedDefault
      : preferredThinkingDefault(levels),
  }
}

export function builtInModelThinking(model: string): ModelThinkingConfig | undefined {
  const id = canonicalThinkingModelID(model)
  const profile = (
    levels: ModelThinkingLevel[],
    defaultLevel: ModelThinkingLevel,
  ): ModelThinkingConfig => ({
    enabled: true,
    levels,
    default_level: defaultLevel,
  })

  if (
    id.includes('claude-fable-5')
    || id.includes('claude-opus-5')
    || id.includes('claude-opus-4-8')
    || id.includes('claude-opus-4-7')
    || id.includes('claude-sonnet-5')
  ) {
    return profile(['low', 'medium', 'high', 'xhigh', 'max'], 'high')
  }
  if (id.includes('claude-opus-4-6') || id.includes('claude-sonnet-4-6')) {
    return profile(['low', 'medium', 'high', 'max'], 'high')
  }
  if (
    id.includes('claude-opus-')
    || id.includes('claude-sonnet-')
    || id.includes('claude-fable-')
  ) {
    return profile(['low', 'medium', 'high'], 'high')
  }

  if (id.includes('gpt-5-6')) {
    return profile(['off', 'low', 'medium', 'high', 'xhigh', 'max'], 'medium')
  }
  if (
    id.includes('gpt-5-5-pro')
    || id.includes('gpt-5-4-pro')
    || id.includes('gpt-5-2-pro')
  ) {
    return profile(['medium', 'high', 'xhigh'], 'high')
  }
  if (id.includes('gpt-5-pro')) return profile(['high'], 'high')
  if (
    id.includes('gpt-5-5')
    || id.includes('gpt-5-4')
    || id.includes('gpt-5-3')
    || id.includes('gpt-5-2')
  ) {
    return profile(['off', 'low', 'medium', 'high', 'xhigh'], 'medium')
  }
  if (id.includes('gpt-5-1')) {
    return profile(['off', 'low', 'medium', 'high'], 'medium')
  }
  if (id.includes('gpt-5') && !id.includes('chat')) {
    return profile(['minimal', 'low', 'medium', 'high'], 'medium')
  }

  const base = id.split('/').at(-1) ?? id
  if (/^o[1-9](?:-|$)/u.test(base)) {
    return profile(['low', 'medium', 'high'], 'medium')
  }
  return undefined
}

function profileFromConfig(
  value: ModelThinkingConfig,
  source: ModelThinkingProfile['source'],
): ModelThinkingProfile {
  const normalized = normalizeModelThinkingConfig(value)
  return normalized.enabled
    ? {
        enabled: true,
        levels: [...normalized.levels],
        defaultLevel: normalized.default_level,
        source,
      }
    : { enabled: false, levels: [], source }
}

function normalizeThinkingLevel(value: unknown): ModelThinkingLevel | undefined {
  const normalized = String(value ?? '').trim().toLowerCase()
  return MODEL_THINKING_LEVELS.find(level => level === normalized)
}

function preferredThinkingDefault(levels: readonly ModelThinkingLevel[]): ModelThinkingLevel {
  return ['medium', 'high', 'low', 'minimal', 'off', 'xhigh', 'max']
    .find((candidate): candidate is ModelThinkingLevel => (
      levels.includes(candidate as ModelThinkingLevel)
    )) ?? levels[0] ?? 'medium'
}

function canonicalThinkingModelID(value: string): string {
  return value.trim().toLowerCase().replaceAll('_', '-').replaceAll('.', '-')
}

File: internal/config/model_thinking.go
package config

import "strings"

var modelThinkingLevelOrder = []string{
	"off",
	"minimal",
	"low",
	"medium",
	"high",
	"xhigh",
	"max",
}

type ModelThinkingConfig struct {
	Enabled      bool     `json:"enabled"`
	Levels       []string `json:"levels,omitempty"`
	DefaultLevel string   `json:"default_level,omitempty"`
}

// ModelThinkingProfile is the bounded, provider-agnostic thinking control
// passed to Pi for one selected model. Pi remains responsible for translating
// the canonical level to the concrete provider request.
type ModelThinkingProfile struct {
	Enabled bool     `json:"enabled"`
	Levels  []string `json:"levels,omitempty"`
	Level   string   `json:"level,omitempty"`
}

func ResolveModelThinking(
	settings AppSettings,
	provider,
	model,
	requested string,
) ModelThinkingProfile {
	provider = strings.TrimSpace(provider)
	model = strings.TrimSpace(model)
	configured, overridden := modelThinkingOverride(settings.ModelThinking, provider, model)
	if !overridden {
		configured, _ = builtInModelThinking(model)
	}
	configured = normalizeModelThinkingConfig(configured)
	if !configured.Enabled || len(configured.Levels) == 0 {
		return ModelThinkingProfile{Enabled: false}
	}
	level := strings.ToLower(strings.TrimSpace(requested))
	if !containsThinkingLevel(configured.Levels, level) {
		level = configured.DefaultLevel
	}
	return ModelThinkingProfile{
		Enabled: true,
		Levels:  append([]string(nil), configured.Levels...),
		Level:   level,
	}
}

func normalizeModelThinkingOverrides(
	value map[string]map[string]ModelThinkingConfig,
	providers map[string]ProviderConfig,
) map[string]map[string]ModelThinkingConfig {
	if len(value) == 0 {
		return nil
	}
	result := make(map[string]map[string]ModelThinkingConfig)
	for rawProvider, models := range value {
		provider := strings.TrimSpace(rawProvider)
		configuredProvider, exists := providers[provider]
		if provider == "" || len(provider) > 64 || (provider != "tokenflux" && (!exists || !configuredProvider.Custom)) {
			continue
		}
		if len(models) > 32 {
			continue
		}
		normalizedModels := make(map[string]ModelThinkingConfig)
		for rawModel, configured := range models {
			model := strings.TrimSpace(rawModel)
			if model == "" || len([]rune(model)) > 256 || strings.ContainsAny(model, "\x00\r\n") {
				continue
			}
			normalizedModels[model] = normalizeModelThinkingConfig(configured)
		}
		if len(normalizedModels) > 0 {
			result[provider] = normalizedModels
		}
	}
	if len(result) == 0 {
		return nil
	}
	return result
}

func normalizeModelThinkingConfig(value ModelThinkingConfig) ModelThinkingConfig {
	selected := make(map[string]bool, len(value.Levels))
	for _, raw := range value.Levels {
		level := strings.ToLower(strings.TrimSpace(raw))
		if validModelThinkingLevel(level) {
			selected[level] = true
		}
	}
	levels := make([]string, 0, len(selected))
	for _, level := range modelThinkingLevelOrder {
		if selected[level] {
			levels = append(levels, level)
		}
	}
	defaultLevel := strings.ToLower(strings.TrimSpace(value.DefaultLevel))
	if !value.Enabled {
		if !containsThinkingLevel(levels, defaultLevel) {
			defaultLevel = preferredThinkingDefault(levels)
		}
		return ModelThinkingConfig{
			Enabled:      false,
			Levels:       levels,
			DefaultLevel: defaultLevel,
		}
	}
	if len(levels) == 0 {
		levels = []string{"low", "medium", "high"}
	}
	if !containsThinkingLevel(levels, defaultLevel) {
		defaultLevel = preferredThinkingDefault(levels)
	}
	return ModelThinkingConfig{
		Enabled:      true,
		Levels:       levels,
		DefaultLevel: defaultLevel,
	}
}

func modelThinkingOverride(
	value map[string]map[string]ModelThinkingConfig,
	provider,
	model string,
) (ModelThinkingConfig, bool) {
	models, exists := value[provider]
	if !exists {
		return ModelThinkingConfig{}, false
	}
	configured, exists := models[model]
	return configured, exists
}

func builtInModelThinking(model string) (ModelThinkingConfig, bool) {
	id := canonicalThinkingModelID(model)
	profile := func(levels []string, defaultLevel string) (ModelThinkingConfig, bool) {
		return ModelThinkingConfig{
			Enabled:      true,
			Levels:       levels,
			DefaultLevel: defaultLevel,
		}, true
	}

	if strings.Contains(id, "claude-fable-5") ||
		strings.Contains(id, "claude-opus-5") ||
		strings.Contains(id, "claude-opus-4-8") ||
		strings.Contains(id, "claude-opus-4-7") ||
		strings.Contains(id, "claude-sonnet-5") {
		return profile([]string{"low", "medium", "high", "xhigh", "max"}, "high")
	}
	if strings.Contains(id, "claude-opus-4-6") ||
		strings.Contains(id, "claude-sonnet-4-6") {
		return profile([]string{"low", "medium", "high", "max"}, "high")
	}
	if strings.Contains(id, "claude-opus-") ||
		strings.Contains(id, "claude-sonnet-") ||
		strings.Contains(id, "claude-fable-") {
		return profile([]string{"low", "medium", "high"}, "high")
	}

	if strings.Contains(id, "gpt-5-6") {
		return profile([]string{"off", "low", "medium", "high", "xhigh", "max"}, "medium")
	}
	if strings.Contains(id, "gpt-5-5-pro") ||
		strings.Contains(id, "gpt-5-4-pro") ||
		strings.Contains(id, "gpt-5-2-pro") {
		return profile([]string{"medium", "high", "xhigh"}, "high")
	}
	if strings.Contains(id, "gpt-5-pro") {
		return profile([]string{"high"}, "high")
	}
	if strings.Contains(id, "gpt-5-5") ||
		strings.Contains(id, "gpt-5-4") ||
		strings.Contains(id, "gpt-5-3") ||
		strings.Contains(id, "gpt-5-2") {
		return profile([]string{"off", "low", "medium", "high", "xhigh"}, "medium")
	}
	if strings.Contains(id, "gpt-5-1") {
		return profile([]string{"off", "low", "medium", "high"}, "medium")
	}
	if strings.Contains(id, "gpt-5") && !strings.Contains(id, "chat") {
		return profile([]string{"minimal", "low", "medium", "high"}, "medium")
	}

	base := id
	if separator := strings.LastIndex(base, "/"); separator >= 0 {
		base = base[separator+1:]
	}
	if len(base) >= 2 && base[0] == 'o' && base[1] >= '1' && base[1] <= '9' &&
		(len(base) == 2 || base[2] == '-') {
		return profile([]string{"low", "medium", "high"}, "medium")
	}
	return ModelThinkingConfig{}, false
}

func canonicalThinkingModelID(value string) string {
	return strings.NewReplacer("_", "-", ".", "-").Replace(strings.ToLower(strings.TrimSpace(value)))
}

func validModelThinkingLevel(value string) bool {
	return containsThinkingLevel(modelThinkingLevelOrder, value)
}

func containsThinkingLevel(levels []string, target string) bool {
	for _, level := range levels {
		if level == target {
			return true
		}
	}
	return false
}

func preferredThinkingDefault(levels []string) string {
	for _, candidate := range []string{"medium", "high", "low", "minimal", "off", "xhigh", "max"} {
		if containsThinkingLevel(levels, candidate) {
			return candidate
		}
	}
	return ""
}
